File: web-sdks/dart/lib/src/models/products/part.dart
import 'package:flutter/material.dart' show Colors, Color;
import 'part_application.dart';
import './item_pricing.dart';
import '../assets/asset.dart';

class Part implements Comparable {
  final int id;
  final String itemType;
  final int sku;
  final int rank;
  final int tier;
  final String mfrCode;
  final String mfrName;
  final String? autoCareBrandId;
  final String partNumber;
  final int? parentPartTypeId;
  final int? partTypeId;
  final String? partTypeName;
  final bool buyersGuide;
  final bool primaryImageExists;
  final bool secondaryImageExists;
  final bool documentExists;
  final bool spinExists;
  final bool attributeExists;
  final bool interchange;
  final String? memberNote;
  final List<PartApplication> applications;
  final String? categoryName;
  final String? subcategoryName;
  Asset? assets;
  ItemPricing? pricing;
  List<Part> equivalents;
  List<Part> alternates;
  List<Part> supercessions;

  Part({
    required this.id,
    required this.itemType,
    required this.sku,
    required this.rank,
    required this.tier,
    required this.mfrCode,
    required this.mfrName,
    this.autoCareBrandId,
    required this.partNumber,
    this.parentPartTypeId,
    this.partTypeId,
    this.partTypeName,
    required this.buyersGuide,
    required this.primaryImageExists,
    required this.secondaryImageExists,
    required this.documentExists,
    required this.spinExists,
    required this.attributeExists,
    required this.interchange,
    this.memberNote,
    required this.applications,
    this.categoryName,
    this.subcategoryName,
    this.assets,
    this.pricing,
    List<Part>? equivalents,
    List<Part>? alternates,
    List<Part>? supercessions,
  }) : equivalents = equivalents ?? [],
       alternates = alternates ?? [],
       supercessions = supercessions ?? [];

  factory Part.fromJson(Map<String, dynamic> json) => Part(
    id: json['id'],
    itemType: json['item_type'],
    sku: json['sku'],
    rank: json['rank'],
    tier: json['tier'],
    mfrCode: json['mfr_code'],
    mfrName: json['mfr_name'],
    autoCareBrandId: json['auto_care_brand_id'],
    partNumber: json['part_number'],
    parentPartTypeId: json['parent_part_type_id'],
    partTypeId: json['part_type_id'],
    partTypeName: json['part_type_name'],
    buyersGuide: json['buyers_guide'],
    primaryImageExists: json['primary_image_exists'],
    secondaryImageExists: json['secondary_image_exists'],
    documentExists: json['document_exists'],
    spinExists: json['spin_exists'],
    attributeExists: json['attribute_exists'],
    interchange: json['interchange'],
    memberNote: json['member_note'],
    applications: ((json['applications'] as List<dynamic>?) ?? [])
        .map((a) => PartApplication.fromJson(a as Map<String, dynamic>))
        .toList(),
    categoryName: json['category_name'],
    subcategoryName: json['subcategory_name'],
    assets: json['asset'] != null
        ? Asset.fromJson(json['asset'] as Map<String, dynamic>)
        : null,
    pricing: json['pricing'] != null
        ? ItemPricing.fromJson(json['pricing'] as Map<String, dynamic>)
        : null,
  );

  @override
  int compareTo(other) {
    int tierCompare = tier.compareTo(other.tier);
    if (tierCompare != 0) return tierCompare;

    int rankCompare = rank.compareTo(other.rank);
    if (rankCompare != 0) return rankCompare;

    int mfrCodeCompare = mfrCode.compareTo(other.mfrCode);
    if (mfrCodeCompare != 0) return mfrCodeCompare;

    int partTypeNameCompare = (partTypeName ?? '').compareTo(
      other.partTypeName ?? '',
    );
    if (partTypeNameCompare != 0) return partTypeNameCompare;

    return partNumber.compareTo(other.partNumber);
  }

  String perCarQuantity() {
    for (final application in applications) {
      for (final display in application.displays) {
        if (display.name == 'Qty') {
          return display.value;
        }
      }
    }
    return "1";
  }

  Color quantityAtLocationColor() {
    Color color = Colors.grey;

    double totalQuantityAvailable = pricing!.locations
        .map((location) => location.quantityAvailable)
        .reduce((value, element) => value + element);

    if (totalQuantityAvailable == 0) return color;

    double quantityAtFirstLocation = pricing!.locations
        .where((location) => location.sortOrder == 1)
        .first
        .quantityAvailable;

    color = quantityAtFirstLocation > 0 ? Colors.green : Colors.amber;

    return color;
  }

  String quantityAtLocationText() {
    double totalQuantityAvailable = pricing!.locations
        .where((location) => location.sortOrder == 1)
        .first
        .quantityAvailable;

    return totalQuantityAvailable > 100
        ? '100+'
        : totalQuantityAvailable.toString();
  }

  double cost() {
    return pricing!.locations
        .where((location) => location.sortOrder == 1)
        .first
        .cost;
  }

  double list() {
    return pricing!.locations
        .where((location) => location.sortOrder == 1)
        .first
        .list;
  }

  double? core() {
    return pricing?.locations
        .where((location) => location.sortOrder == 1)
        .first
        .core;
  }
}

File: web-sdks/javascript/src/models/products/part.js
/**
 * Part model
 */

import { PartApplication } from './part_application.js';
import { ItemPricing } from './item_pricing.js';
import { Asset } from '../assets/asset.js';

/**
 * Part model representing a product part
 */
export class Part {
  constructor({
    id,
    sku,
    rank,
    tier,
    mfrCode,
    mfrName,
    autoCareBrandId = null,
    partNumber,
    parentPartTypeId = null,
    partTypeId = null,
    partTypeName = null,
    buyersGuide,
    productInfo,
    interchange,
    memberNote = null,
    applications,
    assets = null,
    pricing = null,
  }) {
    this.id = id;
    this.sku = sku;
    this.rank = rank;
    this.tier = tier;
    this.mfrCode = mfrCode;
    this.mfrName = mfrName;
    this.autoCareBrandId = autoCareBrandId;
    this.partNumber = partNumber;
    this.parentPartTypeId = parentPartTypeId;
    this.partTypeId = partTypeId;
    this.partTypeName = partTypeName;
    this.buyersGuide = buyersGuide;
    this.productInfo = productInfo;
    this.interchange = interchange;
    this.memberNote = memberNote;
    this.applications = applications;
    this.assets = assets;
    this.pricing = pricing;
  }

  static fromJson(json) {
    return new Part({
      id: json.id,
      sku: json.sku,
      rank: json.rank,
      tier: json.tier,
      mfrCode: json.mfr_code,
      mfrName: json.mfr_name,
      autoCareBrandId: json.auto_care_brand_id,
      partNumber: json.part_number,
      parentPartTypeId: json.parent_part_type_id,
      partTypeId: json.part_type_id,
      partTypeName: json.part_type_name,
      buyersGuide: json.buyers_guide,
      productInfo: json.product_info,
      interchange: json.interchange,
      memberNote: json.member_note,
      applications: (json.applications || []).map(a => PartApplication.fromJson(a)),
      assets: json.asset ? Asset.fromJson(json.asset) : null,
      pricing: json.pricing ? ItemPricing.fromJson(json.pricing) : null,
    });
  }

  /**
   * Compare parts for sorting
   * @param {Part} other
   * @returns {number}
   */
  compareTo(other) {
    let compare = this.tier - other.tier;
    if (compare !== 0) return compare;

    compare = this.rank - other.rank;
    if (compare !== 0) return compare;

    compare = this.mfrCode.localeCompare(other.mfrCode);
    if (compare !== 0) return compare;

    compare = (this.partTypeName || '').localeCompare(other.partTypeName || '');
    if (compare !== 0) return compare;

    return this.partNumber.localeCompare(other.partNumber);
  }

  /**
   * Get per-car quantity from applications
   * @returns {string}
   */
  perCarQuantity() {
    for (const application of this.applications) {
      for (const display of application.displays) {
        if (display.name === 'Qty') {
          return display.value;
        }
      }
    }
    return '1';
  }

  /**
   * Get cost from first location
   * @returns {number}
   */
  cost() {
    if (!this.pricing) return 0;
    const firstLocation = this.pricing.locations.find(loc => loc.sortOrder === 1);
    return firstLocation?.cost || 0;
  }

  /**
   * Get list price from first location
   * @returns {number}
   */
  list() {
    if (!this.pricing) return 0;
    const firstLocation = this.pricing.locations.find(loc => loc.sortOrder === 1);
    return firstLocation?.list || 0;
  }

  /**
   * Get core charge from first location
   * @returns {number|null}
   */
  core() {
    if (!this.pricing) return null;
    const firstLocation = this.pricing.locations.find(loc => loc.sortOrder === 1);
    return firstLocation?.core || null;
  }

  /**
   * Get quantity at first location as text
   * @returns {string}
   */
  quantityAtLocationText() {
    if (!this.pricing) return '0';
    const firstLocation = this.pricing.locations.find(loc => loc.sortOrder === 1);
    const qty = firstLocation?.quantityAvailable || 0;
    return qty > 100 ? '100+' : qty.toString();
  }
}
